### ISSUE TITLE
Changes in RAID behaviour?

### ISSUE THREAD

**Issue by wehimwich** [nvi-inc/fsl11#7]
I am not entirely certain, but I think there may have been some changes in the behaviour of the RAID. Specially, if one uses _drop_primary_ and then reboots, the RAID appears to recover automatically. FSL10 does not seem to do this. Adding use of an `mdadm ... -r ...` before rebooting does not seem to change the result. It also seems like you get this automatic recovery if you boot without the _primary_ disk, shutdown, and then reboot with both disks. However, what I saw may not have been consistent. I will try to test it more. (I am in the process of refreshing my test bed disks in case I have somehow poisoned them.)

If this is correct, I am not sure that it is a big problem, but I think it is going to change some things:

1. If you use _drop_primary_ for "Recoverable testing", you have to hope that the system doesn't get rebooted before you have decided that you have had success. That is a new risk.
3. If you turn the _primary_ disk off for "Recoverable testing" and your testing was successful, you can just reboot with the primary disk turned on to finish (without running _recover_raid_).
4. The discussion of a _missing_ disk in the "recover_raid" script description may need to be revised.
5. If the array has an _inactive_ disk and gets rebooted, it won't be necessary to run _recover_raid_.
6. ...?

**Comment by wehimwich** [comment]
I did some testing relevant to the booting without the _primary_ for testing. I `shutdown`, turned off the _primary_, rebooted, ran and the terminated the FS, shutdown, turned the _primary_ back on and rebooted. The first four times, there was no automatic recovery, so I used _recover_raid_. The fifth time it started recovering automatically. For the next three attempts it did not automatically recover, but the next one after those did. For all of these attempts, it alternated whether it used _/dev/sda1_ or _/dev/sda2_ for _/boot/efi_. On the next attempt, it did not automatically recover, but _recover_raid_ also failed. It seemed to add the _primary_ back in, but then I got the dreaded `ERROR: RAID /dev/md0 failed to start rebuilding`. I mindlessly tried _recover_raid_ a few more times without success. 

Any suggestions for dealing this would be appreciated. My only idea for now is to save the output of `mdadm -D`, `mdadm -E`, `lsblk` and `mdstat` for later analysis and then try to `-f` and `-r` the _primary_ to see if either will help.

**Comment by wehimwich** [comment]
> Any suggestions for dealing this would be appreciated. My only idea for now is to save the output of `mdadm -D`, `mdadm -E`, `lsblk` and `mdstat` for later analysis and then try to `-f` and `-r` the _primary_ to see if either will help.

Neither `-f` or `-r` work with errors `No such device` and `No such device or address` respectively. In _recover_raid_ it is pretty clearly that `mdam /dev/md0 --re-adding missing` is not working, but there is no error output. Trying

````
mdadm /dev/md0 --re-add /dev/sda2
````

yields

````
mdadm --re-add for /dev/sda2 to /dev/md0 is not possible
````

Several files are attached: 
[recover_raid script output](https://github.com/nvi-inc/fsl11/files/9426164/recover_raid.txt)
[mdstat output](https://github.com/nvi-inc/fsl11/files/9426165/mdstat.txt)
[lsblk output](https://github.com/nvi-inc/fsl11/files/9426166/lsblk.txt)
[mdadm -E /dev/sdb2 output](https://github.com/nvi-inc/fsl11/files/9426167/sdb2.txt)
[mdadm -E /dev/sda2 output](https://github.com/nvi-inc/fsl11/files/9426168/sda2.txt)
[mdadm -D /dev/md0 output](https://github.com/nvi-inc/fsl11/files/9426169/md0.txt)

**Comment by jfhquick** [comment]
> > Any suggestions for dealing this would be appreciated. My only idea for now is to save the output of `mdadm -D`, `mdadm -E`, `lsblk` and `mdstat` for later analysis and then try to `-f` and `-r` the _primary_ to see if either will help.
> 
> Neither `-f` or `-r` work with errors `No such device` and `No such device or address` respectively. In _recover_raid_ it is pretty clearly that `mdam /dev/md0 --re-adding missing` is not working, but there is no error output. Trying
> 
> ```
> mdadm /dev/md0 --re-add /dev/sda2
> ```
> 
> yields
> 
> ```
> mdadm --re-add for /dev/sda2 to /dev/md0 is not possible

My suspicion is that you may have mixed up your disks.  The output for sda2 shows it was shed from the array several days before you tried to re-add it?  You could try add a '-v' to the command to see if that gives any more info.  I guess I must have suppressed the output of the re-add for some reason.

If you try using '-a' you should get more success.  The re-add will only work on the disk that was last in the empty slot, but a bare add should happily add anything of a suitable size.

**Comment by wehimwich** [comment]
It is hard to mix up the disks when they were never removed from the machine; the _primary_ was only turned on and off. The reason there was a gap of several days was that I didn't have a chance to get back to it again for awhile. On the original day of testing, this worked eight times in a row: six time with _recovery_raid_, twice with it recovering "automatically", before I hit this snag. I suppose I am saying the obvious, but it seems like the RAID no longer considers the _primary_ to be a member (maybe not the right word) for some reason.

Of course what prompted all this was checking to see if the "automatic" recovery was really happening, which it seems to be, sometimes. Why it is only happens sometimes is a mystery. I have never noticed it happening on FSL10, but maybe I should try this same test with one of those systems.

I tried using `-vv` with `--re-add` and with `-D`, but the output was unchanged.

I tried `-a` and it is now rebuilding. It looks like a full rebuild, rather than then abbreviated rebuild of _recover_raid_. That somewhat defeats the purpose. Dare I trust the result? If so, it is better than swapping disks, running _refresh_secondary_ (and _blank_secondary_ first?), and swapping disks again.  Or maybe I should do that to be safe anyway. Then I suppose I can try this again. :)
